xc/xc7/fasm2bels/make_routes.py

```diff
@@ -303,7 +309,12 @@ def descend_fixed_route(source_node_pkey, fixed_route):
                 if parent_node == self.source_wire_pkey:
                     source_nodes.append(node)
 
-            yield '[list '
+            # TCL cannot express 1-length list, so the curly brackets are
+            # used to prevent TCL from collapsing the 1-length list.
+            if len(source_nodes) == 1:
+                yield '{ '
```

Comment:
Let's just use `'{ '`, I expect it work in all cases.

Reply:
I actually tried this, but I got errors such as the following:

```
CRITICAL WARNING: [Common 17-178] Regular expression error 'Unmatched [ or [^ in character class declaration.  The error occurred while parsing the regular expression fragment: 'GET_NODES$>>>HERE>>>'.' for pattern '[GET_NODES' [/data/symbiflow/symbiflow-arch-defs/build/xc/xc7/tests/soc/litex/mini/minilitex_arty/artix7-xc7a50t-virt-xc7a50t-test/top_bit.v.xdc:9490] 
```

Apparently the `[get_nodes ...]` expression is not compatible with the curly brackets style.

Reply:
Weird :/  Well if this works, great.  Might be worth adding a comment summarizing what you just told me.  It's quiet surprising to me what is and is not working the XDC file, so capturing what doesn't work with a note is useful to future readers of the code.

Reply:
We should revisit this.  I believe we should be able to get everything to work with just using `{` and `}`.  I'm fine merging this as is, but coming back and trying to clean this up is probably worthwhile.

Reply:
From your example below I believe I know how to remove the if statement, and avoid breaking things.

As discussed before, using `{}` causes `[]` expansions to not be done.  This is why when you tried to use `{}` for all lists it didn't work.  But that problem applies to the `len(sources_nodes) == 1` case too!  You just haven't tested a design that hits the problem yet.  I have a solution.

The problem that I was original trying to solve is how to express the following list:

```
a b c { d e } { CLBLM_R_X25Y92/CLBLM_M_C } { g h }
```

Where `CLBLM_R_X25Y92/CLBLM_M_C` is the return value of a `[]` expansion (like get_nodes).

The original problem arises because the behavior of `[list ]` when you send a 1 element list into a list.  This problem can be clearly seen in the following example:
```
set l [list a b c [list d e] [list f] [list g h]]
```
In which case, l becomes `a b c {d e} f {g h}` instead of `a b c {d e} { f } {g h}`.  One solution is a follows:
```
set l [list a b c [format " %s " [list d e]] [format " %s " [list f]] [format " %s " [list g h]]]
```

Which results in: `a b c { d e } { f } { g h }`.

So the real solution is to replace `[list ` with `[format " %s " [list ` and `{} ]` with `]]`.

Reply:
I recommend making to global constants TCL_LIST_OPEN and TCL_LIST_CLOSE, which are 
```
TCL_LIST_OPEN = '[format " %s " [list '
TCL_LIST_CLOSE = ']]'
```

and a block comment explaining why `[list ` and `]` doesn't work as expected.

Reply:
Awesome, tested this solution and it works all right. Hopefully all the tests will not show issues.

I have added also a comment on top of the global constants.